Add tutorial step-97.

cembooks opened the pull request:
Illustrates application of the FE_Nedelec and FE_RaviartThomas finite elements to problems in electromagnetics. Solves the curl-curl equation.

[figures-step-97.tar.gz](https://github.com/user-attachments/files/18577620/figures-step-97.tar.gz)

[step-97.pdf](https://github.com/user-attachments/files/18577618/step-97.pdf)

cembooks commented:
Need help with the test. Step-97 loads meshes from the disk just like step-49 does. When I run the test for step-97 ctest returns the usual deal.II error message informing that the files to be loaded are not found. The test for step-49 runs fine. I need to inform ctest that there are some meshes to be loaded. I do not know how to do that.

marcfehling commented:
> Need help with the test. Step-97 loads meshes from the disk just like step-49 does. When I run the test for step-97 ctest returns the usual deal.II error message informing that the files to be loaded are not found. The test for step-49 runs fine. I need to inform ctest that there are some meshes to be loaded. I do not know how to do that.

Please have a look at the readme in the `tests/examples` folder on how to set up a test for a tutorial:
https://github.com/dealii/dealii/blob/master/tests/examples/README.md

You can have a look at the [`step-49.diff`](https://github.com/dealii/dealii/blob/master/tests/examples/step-49.diff) and [`step-71.diff`](https://github.com/dealii/dealii/blob/master/tests/examples/step-71.diff) files in particular. Those diffs are dealing with relative paths.

cembooks commented:
@marcfehling Thanks! Now I get it.

cembooks commented:
@marcfehling 

I have run "make setup_tests" and "make setup_tests_examples" in the deal.II build directory.  But there are no temporary .cc files to be found at build/tests/examples/source/. Has the directory with the temporary files changed?

![Screenshot from 2025-01-29 10-33-14](https://github.com/user-attachments/assets/95cbf4d6-675e-4218-8801-deea98455c29)

marcfehling commented:
The source file for the test is created at a later stage. I could create it when running the actual test, i.e., with `ctest -R examples/step-97` in your case.

@tamiko We have the `update_diffs` target to update the example tests. However after running the `setup_tests` target, no such target seems to be defined either in my top-level build directory, nor in my `build/tests/examples/source` folder. How do we invoke the target? We should also update the documentation accordingly.

cembooks commented:
@tamiko The same here. Doing "make update_diffs" in the build directory yields "make: *** No rule to make target 'update_diffs'.  Stop."

cembooks commented:
Looks like the "check for doxygen errors and warnings" is the only failing test now. Could you please help me to interpret the log? Can I run the test on the local PC?

cembooks commented:
@marcfehling Thanks! Yes, I am using \f[ in all appropriate paces. Without it MathJax will not generate pictures of equations. It seems to me all equations are ok, see the pdf file attached to the first message. Compiling the doc is not the same as running the test on the server. The doxygen on my local PC does not issue any errors, see the log file attached to this message. 
[doxygen.log](https://github.com/user-attachments/files/18601910/doxygen.log)

marcfehling commented:
> Thanks! Yes, I am using \f[ in all appropriate paces.

@cembooks You are using it everywhere except for the one place that I highlighted :)

cembooks commented:
@marcfehling 
Thanks for the help! Now all tests run ok.

cembooks commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,937 @@
+
+@htmlonly
+<p align="center">
+  <img src="images/steps/developer/step-97-coil.svg" alt="The coil" height="425">
```
Last time I have seen this image in figures-step-97.tar.gz.
It is also attached to this message. Can you open it?

![step-97-coil](https://github.com/user-attachments/assets/d2eec1a6-5364-48c5-92fa-7d10e90f98c9)

cembooks replied:
The tarball figures-step-97.tar.gz is attached to the first message above.

bangerth reviewed (commented):
Hi @cembooks : First, many thanks for this program! I meant to get through more of it today, but then got stuck in meetings. But it's the summer break and I think I will be able to read through the rest of the program over the next couple of weeks. In the meantime, here are a few comments already.

Like I said, many thanks -- we'll see that we get that merged before the next release if at all possible!

cembooks commented on examples/CMakeLists.txt:
```diff
@@ -44,6 +44,14 @@ if(DEAL_II_COMPONENT_EXAMPLES)
     PATTERN "topography.txt.gz"                 # step-53
     PATTERN "input/initial_mesh_3d.vtk"         # step-54
     PATTERN "input/DTMB-5415_bulbous_bow.iges"  # step-54
+    PATTERN "sphere_r6.geo"                     # step-97
+    PATTERN "sphere_r6.msh"                     # step-97
+    PATTERN "sphere_r7.geo"                     # step-97
+    PATTERN "sphere_r7.msh"                     # step-97
+    PATTERN "sphere_r8.geo"                     # step-97
+    PATTERN "sphere_r8.msh"                     # step-97
+    PATTERN "sphere_r9.geo"                     # step-97
+    PATTERN "sphere_r9.msh"                     # step-97
```
@bangerth Thanks for checking the tutorial!!! 

I use them all for creating the convergence tables.  The section “Results” contains two convergence tables. The second column of each table contains the mesh refinement parameter, ‘r = 6,7,8,9’. Each value of the parameter is associated  with a mesh file.

bangerth replied:
Great, let's keep all these files then!

bangerth replied:
Perhaps worth noting: None of these files are particularly large (~30,000 lines of text).

cembooks replied:
This makes me thinking...

I have measured the disk space used by all step folders on my local PC. All step folders (from step-1 to step-97) use 15.4MB. The *.msh files of the step-97 tutorial use 5.4MB. That is, one tutorial (out of almost hundred) uses about 1/3 of the total disk space. This seems to be somewhat silly. I guess this needs to be fixed...

In my library (https://github.com/cembooks/static) I store only one *.geo file and use a bash script to generate multiple *.msh files at build time. Maybe we can do a similar thing in here as well: keep only one *.geo file and generate *.msh files at the build or run time. Maybe we can use some methods of step-49 tutorial. That is, construct the mesh inside the deal.II code. I need some more time to consider the options. Any advise is welcome!

Can we go on reviewing the rest of the tutorial while I am rethinking the mesh issue? I guess fixing the mesh issue will affect only a small portion of the code.

cembooks commented on doc/doxygen/tutorial/tutorial.h.in:
```diff
@@ -1369,7 +1377,13 @@
+ *     <td> The curl-curl equation
+ *     </td>
+ *     <td>step-97</td>
+ *   </tr>
```
step-81 added.

cembooks commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,937 @@
+The outer shell, it is depicted in blue color,
+contains free current. The current fills all the space available in the wall of
+the shell. In the figure below, however, only three cross sections of the
+current are shown. The current density is the highest on the equator and equals
```
By “free current” I mean the following. There is the “total current” on the right-hand side of the Ampere’s law of the Maxwell’s equations. In the static approximation the “total current” is a sum of “free current” and “bound current” as the “displacement current” equals zero. The “free current” literally means “not bound current”. The “free current” is charge moving in space. The “bound current” is the net effect of magnetization. In physical models the “bound current” can replace magnetization. A magnetic field can be induced by a current-caring coil, i.e., by the “free current”, as well as by a piece of permanent magnet, i.e., by the “bound current”. That is, both “free current” and “bound current” can, in general, be “prescribed current”.

I have the following text suggestion:

“The outer shell, depicted in blue, contains the current-carrying windings of the coil. The windings are modeled by a prescribed free-current density. The prescribed free-current density fills all the space available in the wall of the shell. In the figure below, however, only three cross sections of the free-current density are shown. The free-current density is the highest on the equator and equals zero at the two poles. It can be described as ...”

bangerth replied:
Yes, this actually reads pretty well -- let's go with that!

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+Neumann boundary conditions as it allows to reach the same level of simulation
+error with a sphere $\Gamma_{R1}$ of a smaller radius.
+
+All three boundary conditions, Dirichlet, Neumann, and Robin, guaranty the
```
```suggestion
All three boundary conditions, Dirichlet, Neumann, and Robin, guarantee the
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+uniqueness of the solution itself, $\vec{A}$, can only be guaranteed if
+one of the boundary conditions is applied in combination with a gauge, see the
+discussion above. The program uses the Robin boundary condition by default,
+but can by easily switched to Dirichlet or Robin boundary conditions.
```
```suggestion
but can easily be switched to Dirichlet or Robin boundary conditions.
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+interface. No free currents flow on the surface of the magnetic core of the coil
+described above. Therefore, we set $\vec{K}_f = 0$. Then by substituting the
+constitutive relation for the magnetic field and the definition of the magnetic
+vector potential, see above, into last two equations we get
```
```suggestion
vector potential, see above, into the last two equations we get
```

bangerth commented on examples/step-97/doc/kind:
```diff
@@ -0,0 +1 @@
+basic
```
I think this might fit better into the "techniques" category. Would you mind changing this?

cembooks replied:
Done.

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+norms at these stages makes no sense as we apply implicit gauges and the
+conservative portion of the solution is unknown. We can, however, judge the
+quality of the simulations at these two stages indirectly by evaluating the
+quality of the simulations at the second and the fourth stages.
```
938 lines for the introduction makes this program effectively tied for third-longest (step-32 has 1300, step-31 has 1020, step-22 has 950). Thank you for writing so eloquently and comprehensively!

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+current context as well.
+
+The current vector potential is computed in free space, i.e., $\mu = \mu_0$
+everywhere. For this reason, there is no need in interface conditions as
```
```suggestion
everywhere. For this reason, there is no need for interface conditions as
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+It is beneficial to apply the homogeneous Dirichlet boundary condition as it
+allows to nullify the integral $I_{b3-2}$ in the recipe for calculating the
+magnetic vector potential, see below. Consequently, we can neglect this
+integral when programming what makes the code more efficient.
```
```suggestion
integral when programming and this makes the code more efficient.
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+Any magnetic vector potential that minimizes this functional will satisfy
+the curl-curl equation (i), the boundary condition (ii) and the interface
+condition (iv). On the other hand, this functional is invariant to the
+interface condition (iii). This can be verified as the following. The
```
```suggestion
interface condition (iii). This can be verified as follows: The
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+  boundary value problem above with equation (ii) replaced by the Dirichlet
+  boundary condition.
+
+In general, we need to sort all boundary and interface conditions into two
+categories, natural and essential. The reason for doing so is simple: we need to
```
Somewhere in this section, would you mind referencing step-22 and the longish discussion of boundary values there?

cembooks replied:
Done.

cembooks replied:
I cite the video lectures by using html links. Is there a better way to cite the lectures? 
I have tried to use the same links as in step-22, that is, “@dealiiVideoLecture{21.5}”.  But then the “indent/doxygen” test failed so I had to resort to html links.

bangerth replied:
Do it like in the other tutorial programs, using `@dealiiVideoLecture` and we'll sort that out later on.

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+magnetic vector potential computed by the finite element method, is represented
+as a sum,
+\f[
+\vec{A} = \sum_{j=0}^{m-1} c_j \vec{N}_j,
```
Let's make clear which of these are functions and which are just numbers:
```suggestion
\vec{A}(\vec{x}) = \sum_{j=0}^{m-1} c_j \vec{N}_j(\vec{x}),
```

cembooks replied:
I think it is better to use $\vec{r}$ to be consistent with the choice of the coordinate system.

bangerth replied:
I'm happy to leave that up to you -- as long as it is consistent within the introduction, all I wanted to point out is that adding the argument makes it clear that `c` are the coefficients and `N` are the shape functions.

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+
+<h4>Converting potentials into fields</h4>
+
+After the magnetic vector potential is computed, it needs to be converted into
+the magnetic field as
+\f[
+\vec{B} = \vec{\nabla} \times \vec{A}.
+\f]
```
You don't *need* to, but this step is *useful* because what we are interested in for physical reasons is B. Perhaps say that?

cembooks replied:
"After the magnetic vector potential is computed, it needs to be converted into
the magnetic field as"

is replaced by

"As discussed above, we are interested in the magnetic field induced by the coil.
For this reason, we convert the numerically computed vector potential into magnetic
field as"

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+\f[
+\vec{J}_f = \vec{\nabla} \times \vec{T}.
+\f]
+This equation is, essentially, the same as the equation for the converting
```
```suggestion
This equation is, essentially, the same as the equation for converting
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+The preceding sections leave an impression that we are going to model $\vec{A}$
+and $\vec{T}$ by the FE_Nedelec finite elements, while $\vec{B}$ and $\vec{J}_f$
+by the FE_RaviartThomas finite elements. It makes sense to discuss how this
+kind of choices can be made.
```
```suggestion
kind of choice can be made.
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+by the FE_RaviartThomas finite elements. It makes sense to discuss how this
+kind of choices can be made.
+
+The most simple method of assigning a particular type of finite elements to a
```
```suggestion
The simplest method of assigning a particular type of finite elements to a
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+The most simple method of assigning a particular type of finite elements to a
+physical quantity studied in electromagnetics is a contemplation of the Bossavit's
+diagram. It is presented below. The Bossavit's diagram describes the relations
+between physical quantities as they are given by the Maxwell's equations and
+constitutive relations. Most equations derived from the Maxwell's equations
```
It's either "the Maxwell equations" or "Maxwell's equations". Either is fine, but not "the Maxwell's equations".
```suggestion
between physical quantities as they are given by Maxwell's equations and
constitutive relations. Most equations derived from Maxwell's equations
```

bangerth commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+The program uses four meshes created with a help of gmsh. Each mesh has a
+different degree of refinement. The degree of refinement in each mesh is
+defined by a number of mesh nodes on the transfinite lines, $r$. The figure
+below illustrates a mesh with $r=5$. The program uses for meshes with the
```
```suggestion
below illustrates a mesh with $r=5$. The program uses four meshes with the
```

bangerth reviewed (commented):
@cembooks I read through the introduction and greatly enjoyed that -- thank you for writing extensively and clearly! Almost all of the comments I have are just language issues.

Do you reference your book somewhere? If you don't, I think it would be entirely appropriate to do so. You can put an entry into `doc/doxygen/references.bib` and reference it via `@cite`.

I'll get to the rest of the program in the next few days. In the meantime, please email me the pictures that accompany the program so I can put them into the website repository. Alternatively, open a pull request similar to https://github.com/dealii/website-large_assets/pull/7, using the same naming convention as used there.

bangerth commented on examples/step-97/doc/results.dox:
```diff
@@ -0,0 +1,116 @@
+
+The program also dumps a number of files in the current directory. In the default
+configuration these files are:
+- vtu files. They contain the computed vector fields. Recall that the program
+  uses the second-order mapping. For this reason, one needs to use a fresh
+  version of ParaView to open them. Visit will not do.
```
I imagine that is because you are outputting cells with curved edges? Perhaps say that, and link to the discussion in `DataOut` or `DataOutBase` where er talk about this? (Assuming you're not already doing that in the .cc file -- I haven't looked at that yet.)

cembooks replied:
I have found this wiki page:
https://github.com/dealii/dealii/wiki/Notes-on-visualizing-high-order-output
It has been updated on Dec 13, 2024.  I think it is better to cite this page. It has section "Extra details" with references to the documentation and tutorials.

bangerth replied:
Yes, that sounds good.

bangerth commented on examples/step-97/doc/results.dox:
```diff
@@ -0,0 +1,116 @@
+ParaView has "Plot Over Line" filter. You can use this filter to visualize the
+difference between the exact solution and a solution simulated with a particular
+boundary condition. You can also draw conclusions by observing the convergence
+tables. Keep in mind the $\eta^2$ parameter. Increase it if the CG solver chocks
```
```suggestion
tables. Keep in mind the $\eta^2$ parameter. Increase it if the CG solver chokes
```

bangerth commented on examples/step-97/doc/results.dox:
```diff
@@ -0,0 +1,116 @@
+boundary condition are observed the best when higher-order finite elements are
+used, i.e., $p = 1$ and $p = 2$.
+
+The Robin boundary condition as described above is also called the first-order
+asymptotic boundary condition (ABC). There exist ABCs of higher orders. Implement
+and test the second-order ABC to see if it performs any better. There exist
+improvised asymptotic boundary conditions, IABCs. Try to implement the first
+order IABC.
```
If you have a reference, this would be a good place to put it to help people find out more about these ABCs.

bangerth reviewed (commented):
Here's the results section.

cembooks commented:
@bangerth Thank you for checking the tutorial! I have sent you the pictures that accompany the program via e-mail.

cembooks commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+\vec{\nabla}\times\big(\vec{C} + \vec{S} \big) =
+\vec{\nabla}\times\vec{S},
+\f]
+we are interested in the solenoidal part of the solution. What the conservative
+part of the solution is, exactly, not important. We need to gauge because we
+need to select one solution out of many. Which conservative component of the
+solution is brought by the gauge is not important.
```
I agree. We can improve this paragraph. The really sad thing is that the magnetic field, B, is not a physical reality. Only the electric field, E, is. The magnetic field is a trick played on us by the relativity and charge invarince, see, for instance, section 12.3.1 in ‘Introduction to electrodynamics’ by David J. Griﬃths, https://archive.org/details/davidgriffithselectrodynamics/page/n537/mode/2up. Even the magnetic field cannot be trusted these days – it is just an illusion. Although, the magnetic field, unlike the vector potential, is measurable.

I have the following text suggestion:

“ ...
speciﬁes that there is no conservative component in A. That is, the Coulomb gauge selects a purely solenoidal solution. We are interested in the magnetic ﬁeld, B. In the realm of classical electromagnetics, the magnetic vector potential, A, has an instrumental value - we just use it to compute the magnetic field, B. The magnetic vector potential is a usefull mathematical tool, but it has no physical meaning as it cannot be measured. On the contrary, the magentic field, B, can be measured and, as a consequence, has a distinctive physical meaning. From a physical perspective, there are (infinitely) many magnetic vector potentials that lead to the same physically measurable magnetic field and for a physicist, it doesn't matter which of these potentials we choose. But from a mathematical perspective, we need to pick one to make the problem well-posed. A gauge is a way to pick one. Because the magnetic ﬁeld is computed as
\f[
\vec{B} = \vec{\nabla}\times\vec{A} =
\vec{\nabla}\times\big(\vec{C} + \vec{S} \big) =
\vec{\nabla}\times\vec{S},
\f]
we are interested in the solenoidal part of the solution.  Which conservative component of the solution is brought by the gauge is not important.
…”

bangerth replied:
Yes, sure, that looks good to me. (But replace "usefull" by "useful".)

I think the discussion about relativistic perspectives on the magnetic field being an illusion is perhaps outside the scope of your program -- though it would be fun to eventually see a program that solves everything with 4-vectors :-)

cembooks commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+Equation (i) in the boundary value problem has been discussed above and needs
+no further comments.
+
+The surface $\Gamma_{R1}$ represents infinity. The magnetic field induced by
```
"...
The surface $\Gamma_{R1}$ represents infinity. The magnetic field induced by
the coil vanishes at infinity, so does the magnetic vector potential. For this
reason, the homogeneous Dirichlet,
..."

is replaced by

"...
﻿The surface $\Gamma_{R1}$ represents infinity. As most problems in magnetics,
the problem we would like to solve is unbounded. That is to say, the magnetic
field as well as the magnetic vector potential extend to infinity. A
straightforward modeling of an unbounded domain by means of finite elements is
impossible as such modeling will require an infinite amount of cells. To
overcome this unfortunate predicament, we artificially truncate the unbounded
problem domain with the surface $\Gamma_{R1}$. This truncation will introduce a
simulation error. We, however, will try to minimize this error by making the
interior of the artificial surface $\Gamma_{R1}$ as spacious as possible.
The magnetic field induced by the coil vanishes at infinity, so does the
magnetic vector potential. For this reason, the homogeneous Dirichlet,
..."

bangerth replied:
I like that you say "unfortunate predicament" :-)

cembooks commented on examples/step-97/doc/intro.dox:
```diff
@@ -0,0 +1,938 @@
+(assume $\eta^2 = 0$ for a moment). Therefore, the two sides of the equation
+will be somewhat incompatible: pure solenoidal vector field on the left-hand
+side and a combination of a conservative and solenoidal vector fields on the
+right-hand side. To make them compatible we derive the free-current density
```
I agree. However, I think that words “range” and “operator” can discourage many readers in electromagnetics. 

I have the following text suggestion:

"...
Therefore, the two sides of the equation
will be somewhat incompatible: pure solenoidal vector field on the left-hand
side and a combination of a conservative and solenoidal vector fields on the
right-hand side. As a consequence, we cannot expect that we can get the norm
of the residual to zero. In other words, the linear solver cannot converge
to a solution that has zero residual. In practice, this means that CG will
either not converge at all, or terminate only after an unreasonably large
number of iterations. To make the two sides of the equation compatible, we derive
the free-current density from a current vector potential, $\vec{T}$, as
\f[
\vec{J}_f = \vec{\nabla}\times\vec{T}.
\f]
Consequently, the curl-curl equation becomes
\f[
\vec{\nabla}\times\bigg(\dfrac{1}{\mu} \vec{\nabla}\times\vec{A}\bigg) +
\eta^2\vec{A} = \vec{\nabla}\times\vec{T}.
\f]
This equation contains purely solenoidal vector fields on both sides if $\eta^2=0$.
Now, however, we need to convert $\vec{J}_f$ into $\vec{T}$. 
..."

bangerth replied:
Yes, that sounds great.

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+  }
+
+private:
+  int                      dimensions;
+  std::vector<std::string> new_order = {"p", "r", "ncells", "ndofs", "L2"};
```
```suggestion
  const unsigned int                      dimensions;
  std::vector<std::string> new_order = {"p", "r", "ncells", "ndofs", "L2"};
```

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+  // This function describes the free-current density, $\vec{J}_f$, inside the
+  // current region.
+  inline Tensor<1, 3>
+  volume_free_current_density(double x, double y, double z, double K0)
```
Here and everywhere, would you mind following our usual convention to mark *input* arguments as `const`? We do that universally; moreover, if the type of the input argument is bigger than a scalar (say, a `Tensor<1,dim>`), then we also mark it as a reference -- but that is not necessary here.

cembooks replied:
Done with exceptions of functions that belong to WorkStream.

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+                                          double a2,
+                                          double b2)
+  {
+    double cos_theta;
+    double sin_theta;
+
+    double cos_phi;
+    double sin_phi;
+
+    Tensor<1, 3> r_hat;
+    Tensor<1, 3> theta_hat;
+
+    Tensor<1, 3> F1;
+    Tensor<1, 3> F2;
+    Tensor<1, 3> Bj;
+
+    double r;
+
+    r = sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));
+
+    cos_theta = z / r;
+    sin_theta = sqrt(pow(x, 2) + pow(y, 2)) / r;
+
+    cos_phi = x / sqrt(pow(x, 2) + pow(y, 2));
+    sin_phi = y / sqrt(pow(x, 2) + pow(y, 2));
+
+    r_hat[0] = x / r;
+    r_hat[1] = y / r;
+    r_hat[2] = z / r;
+
+    theta_hat[0] = cos_theta * cos_phi;
+    theta_hat[1] = cos_theta * sin_phi;
+    theta_hat[2] = -sin_theta;
```
This is C style :-) You're declaring (and initializing!) the variables at the top, and then you're setting their values below. C++ style would move the declaration as far down as you can, which would yield something like this:
```suggestion
    const double r = sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));

    const double cos_theta = z / r;
    const double sin_theta = sqrt(pow(x, 2) + pow(y, 2)) / r;

    const double cos_phi = x / sqrt(pow(x, 2) + pow(y, 2));
    const double sin_phi = y / sqrt(pow(x, 2) + pow(y, 2));

    const Tensor<1, 3> r_hat = { x / r, y / r, z/r };
    const Tensor<1,3> theta_hat = { cos_theta * cos_phi, cos_theta * sin_phi, -sin_theta};
```
and similar for `F1, F2, Bj`.

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+
+      for (unsigned int i = 0; i < values.size(); i++)
+        {
+          r = sqrt(pow(p[i][0], 2) + pow(p[i][1], 2) + pow(p[i][2], 2));
```
In general, qualify names like this:
```suggestion
          r = std::sqrt(std::pow(p[i][0], 2) + std::pow(p[i][1], 2) + std::pow(p[i][2], 2));
```
But in the specific case here, this is simpler:
```suggestion
          r = p[i].norm();
```

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+
+          if ((r >= a2) && (r <= b2))
+            {
+              Jf = volume_free_current_density(p[i][0], p[i][1], p[i][2], K0);
```
This would suggest that a better design would perhaps let `volume_free_current_density()` take a `Point<3>` as first argument, rather than separate x, y, and z-values?

cembooks replied:
Done.

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+
+      for (unsigned int i = 0; i < values.size(); i++)
+        {
+          B = magnetic_field_coil(p[i][0], p[i][1], p[i][2], K0, mu_0, a2, b2) +
```
same here with the separate x/y/z/ arguments, and of course also in the next line

cembooks replied:
Done.

bangerth commented:
Separately, github has become *quite* slow on this pull request. Would you mind squashing everything into one or a small number of commits?

cembooks commented:
I have tried to do squash-rebase using the same commands I used with the last PR. This time something went wrong. PR step-97 now has 582 commits. Could you please help me to squash-rebase?

bangerth commented:
Oh dear! First step: make a backup of your files by doing
```
git checkout step_97
git branch step_97_backup    # creates a new branch at the current commit
```
This way you have a branch that points to your current last commit in case anything goes wrong.

What happens if you then do the following:
```
git checkout master
git pull upstream master
git checkout step_97
git rebase master
```
Ideally, if you do `git log`, it should only show your commits.

If that worked, then do
```
git push -f origin step_97
```
and we can think about the next steps.

bangerth commented:
For now, do `git rebase --abort`. Do you have any pending changes locally on your hard drive? If not, let me know and I'll fix this for you :-)

cembooks commented:
> For now, do `git rebase --abort`. Do you have any pending changes locally on your hard drive? If not, let me know and I'll fix this for you :-)

I have done git rebase –abort. I do not have any  pending changes. Could you please fix this for me?

cembooks commented:
> @cembooks I don't know how exactly you ended up in the state you are in, but it doesn't matter -- I have a clean branch for you :-) I even merged all commits into one.
> 
> Here's what you should do:
> 
>     * I assume you've already created the backup `step_97_backup` branch as described above
> 
>     * Delete the existing branch:
> 
> 
> ```
> git checkout master
> git branch -D step_97
> ```
> 
>     * Add my repository to yours as a remote:
> 
> 
> ```
> git remote add bangerth git@github.com:bangerth/dealii.git
> ```
> 
>     * Get my merged branch:
> 
> 
> ```
> git fetch bangerth step_97
> ```
> 
>     * Check it out (this will become your new `step_97` branch):
> 
> 
> ```
> git checkout step_97
> ```
> 
> Look around and make sure that it contains everything you think should be on that branch. If you like what you see, you can do
> 
> ```
> git push -f origin step_97
> ```
> 
> to update what is in the pull request.
> 
> Let me know if that works!


Thanks! It is working now! 

The command “git checkout step_97” failed. But I did “git checkout --track bangerth/step_97” instead. Git suggested this. I hope this is not a problem …

I have attached the screenshots in case we want to know what got me into trouble this time if there is any trouble. 

![Screenshot - 1](https://github.com/user-attachments/assets/003c7c1d-391a-425f-941f-4eece3a997d8)
![Screenshot - 2](https://github.com/user-attachments/assets/5f3d9989-e700-4585-9711-14553840c717)

cembooks commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+#include <string>
+#include <ostream>
+
+#define VE scratch_data.ve
+
+#define TMR(__name) TimerOutput::Scope timer_section(timer, __name)
```
Done.

cembooks commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+// solutions. If the projected exact solution and the numerical solution look
+// alike, it is a good sign. In case of problems with the convergence of the
+// CG solver the parameter $\eta^2$ must be increased just a bit.
+class Settings
+{
+public:
+  Settings(){};
+
+  const double permeability_fs = 1.2566370614359172954e-6;
+  const double mu_0            = permeability_fs; // Permeability of free space.
+
+  const double mu_r = 4; // Relative permeability of the magnetic material.
+  const double mu_1 = mu_0 * mu_r; // Permeability of the magnetic material.
+  const double d1   = 0.1; // Half-side of the cube in the center of the mesh.
+  const double a1   = 0.3; // Inner radius of the magnetic core.
+  const double b1   = 0.6; // Outer radius of the magnetic core.
+  const double a2   = 0.9; // Inner radius of the free-current region.
+  const double b2   = 1.2; // Outer radius of the free-current region.
+  const double d2   = 2.0; // Error norms are computed if r < d2.
+  const double d3 = 3.0; // Radius of the outer boundary of the problem domain.
+  const double K0 = 1.0; // Magnitude of the free-current density.
+  const double H0 =      // Magnitude of H-field inside the coil, core removed.
+    (1.0 / 3.0) * K0 * (pow(b2, 2) - pow(a2, 2));
+  const types::material_id mid_1 = 1; // Material ID of the free space.
+  const types::material_id mid_2 = 2; // Material ID of the magnetic core.
+  const types::material_id mid_3 = 3; // Material ID of the free-current region.
+  const types::boundary_id bid   = 2; // Boundary ID. Set in the geo file.
+
+  const unsigned int          fe_degree = 0; // Degree of the finite elements.
+  const BoundaryConditionType boundary_condition_type_A = Robin;
+
+  const double eta_squared_T = 0.0; // eta^2 when solving for T$.
+  const double eta_squared_A = 0.0; // eta^2 when solving for A$.
+
+  const unsigned int nr_threads_max = 0; // If >0 limits the number of threads.
+  const double eps = 1e-12; // Two doubles are equal if their difference < eps.
+  const bool   log_cg_convergence = false; // Save the CG convergence data.
+  const bool print_time_tables = false; // Print the time tables on the screen.
+  const bool project_exact_solution = false; // Save the exact solution.
+};
```
Done.

cembooks commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2170 @@
+
+// Next comes the weight function used to limit the region in which the $L^2$
+// error norms are computed. The error norms are computed if $r < d2$.
+class Weight : public Function<3>, private Settings
+{
+public:
+  virtual double value(const Point<3> &p,
+                       const unsigned int) const override final
+  {
+    if (p.norm() > d2)
```
Done. I guess I need to correct this in my library as well. I use class derivation to distribute the settings to all relevant classes. I thought it was handy. I did not realize that people think in terms of “is a” relationship although I read about it in the books.

bangerth replied:
I acknowledge that it's *convenient* to do it via derivation, but I think there is great value in following established approaches such as "inheritance = is-a relationship" and "member variable = has-a relationship" simply because it's a concept widely taught and understood. One of the things that I think makes the tutorials so useful is that they all attempt to follow very similar styles and that once you've understood the style, you can focus on the content -- in other words, we purposefully try to make all of these programs look very similar. One of my main regrets about some of the earlier tutorials (and in particular step-13 and step-14) is that I had not understood the value of similarity at the time, and these programs are difficult to follow because they show what you can do with C++, not because it is necessary for that program but because it is possible, and I would consequently characterize these programs as "form over function".

bangerth commented:
I'm glad you got it to work. I will be looking at the remainder of the program in the next few days. In the meantime:

* The system currently reports the following error:
```
/home/runner/work/dealii/dealii/build/tests/examples/source/step-97.cc:107:16: error: unused variable 'd3' [-Werror,-Wunused-const-variable]
  107 |   const double d3 = 3.0; // Radius of the outer boundary of the problem domain.
      |                ^~
```
* When you address a comment, and there is nothing that needs to be discussed, there is no need to say "Done" or similar. I get an email for each "Done" :-)

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2125 @@
+            // The variable `curl_vec_in_Hcurl` denotes the curl of the input
+            // vector field, $\vec{\nabla} \times \vec{T}$ or
+            // $\vec{\nabla} \times \vec{A}$, depending on the context.
+            scratch_data.curl_vec_in_Hcurl[0] =
+              scratch_data.vector_gradients.at(q_index).at(2)[1] -
+              scratch_data.vector_gradients.at(q_index).at(1)[2];
+
+            scratch_data.curl_vec_in_Hcurl[1] =
+              scratch_data.vector_gradients.at(q_index).at(0)[2] -
+              scratch_data.vector_gradients.at(q_index).at(2)[0];
+
+            scratch_data.curl_vec_in_Hcurl[2] =
+              scratch_data.vector_gradients.at(q_index).at(1)[0] -
+              scratch_data.vector_gradients.at(q_index).at(0)[1];
```
You're extracting the curl from the tensor gradients here. Couldn't you get these directly if you replaced `scratch_data.fe_values_Hcurl.get_function_gradients` by `scratch_data.fe_values_Hcurl[ve].get_function_curls`?

Separately: Instead of `.at(q_index).at(...)`, let's use `[q_index][...]` as you do everywhere else (I believe).

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2125 @@
+  // solution.
+  void Solver::compute_error_norms()
+  {
+    Weight                     weight;
```
```suggestion
    const Weight                     weight;
```

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2125 @@
+      {
+        const std::vector<double> history_data = control.get_history_data();
+
+        std::ofstream ofs(fname + "_cg_convergence.csv");
+
+        unsigned int i = 1;
+        for (auto item : history_data)
+          {
+            ofs << i << ", " << item << "\n";
+            i++;
+          }
+        ofs.close();
```
Same comments here as above.

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2125 @@
+
+    std::ofstream ofs;
+
+    DataOutBase::VtkFlags flags;
+    flags.write_higher_order_cells = true;
+    data_out.set_flags(flags);
+
+    const MappingQ<3> mapping(mapping_degree);
+
+    data_out.build_patches(mapping,
+                           fe_Hdiv.degree + 2,
+                           DataOut<3>::CurvedCellRegion::curved_inner_cells);
+
+    ofs.open(fname + ".vtu");
+    data_out.write_vtu(ofs);
+
+    ofs.close();
```
and here

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2125 @@
+public:
+  void run()
+  {
+    if (Settings::nr_threads_max)
```
Using `nr_` as a prefix is not common in English. Perhaps use `n_` instead?

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2125 @@
+// @sect3{The main loop}
+
+// This class contains the main loop of the program.
+class Batch
```
It's not clear to me what `Batch` here refers to. Perhaps choose a different name? Maybe `MagneticProblem` or `MagnetoStaticProblem`? I think that's roughly in line with what other programs typically use.

bangerth commented on examples/step-97/step-97.cc:
```diff
@@ -0,0 +1,2125 @@
+    if (Settings::nr_threads_max)
+      MultithreadInfo::set_thread_limit(Settings::nr_threads_max);
+
+    const unsigned int m = 2; // Mapping degree.
+    const unsigned int p = Settings::fe_degree;
```
Since you already have `Settings::fe_degree`, would it be useful to also add `Settings::mapping_degree`?

cembooks commented:
> Here is the rest -- this is all just minor stuff, I think the design of the program is well done.

Thank you for your time and help! 

I have adopted the code suggestions. I still need to add some references to intro.dox and references.bib. I will squash-rebase and then add the references if rebasing will not lead me into a trouble again.

cembooks commented:
> Sounds good. Let us know when you're done and/or need help!

I have added the references. The code with referenced has passed all the tests. Then I have rewritten the first few lines of the SolverT::Solver::make_mesh() so it can handle the compressed mesh files. I have just copy-pasted the code from step-53. Now I have a ctest error:

Error: Process completed with exit code 1.

On my local PC ctest runs just fine. I cannot see any other information regarding the error. Could you please help me to pass this test?

bangerth commented:
I think this is because of the following linker errors:
```
[D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
step-97.obj : error LNK2001: unresolved external symbol "int const boost::iostreams::zlib::default_compression" (?default_compression@zlib@iostreams@boost@@3HB) [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
step-97.obj : error LNK2001: unresolved external symbol "int const boost::iostreams::zlib::deflated" (?deflated@zlib@iostreams@boost@@3HB) [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
step-97.obj : error LNK2001: unresolved external symbol "int const boost::iostreams::zlib::default_strategy" (?default_strategy@zlib@iostreams@boost@@3HB) [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
step-97.obj : error LNK2001: unresolved external symbol "int const boost::iostreams::zlib::okay" (?okay@zlib@iostreams@boost@@3HB) [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
step-97.obj : error LNK2001: unresolved external symbol "int const boost::iostreams::zlib::stream_end" (?stream_end@zlib@iostreams@boost@@3HB) [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
step-97.obj : error LNK2001: unresolved external symbol "int const boost::iostreams::zlib::sync_flush" (?sync_flush@zlib@iostreams@boost@@3HB) [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
D:\a\dealii\dealii\build\tests\examples\binary\step-97.debug\Debug\step-97.debug.exe : fatal error LNK1120: 18 unresolved externals [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-97.debug.vcxproj]
```
I imagine that you have to add these lines to your `CMakeLists.txt` file:
https://github.com/dealii/dealii/blob/master/examples/step-53/CMakeLists.txt#L37-L48

cembooks commented:
> That's ok with me. Go ahead and undo the change then. We can always play with this as a follow-up if we want.

I have removed the file compression and looked at the tutorial on my local computer. Everything looks fine to me. The tests on the server run without errors. In my opinion we can merge the tutorial if you do not have any corrections or suggestions.

cembooks commented:
> The current state is fine with me. Would you mind squashing your 8 commits into one once again? Or would you like me to do that?

Thanks! After saying my prayers I have copied the git magic words from the logbook into LCI. It looks like the squash-rebase was successful this time.

I have another question. The PETSc has access to Hypre AMS. The deal.II has PETSc wrappers. I was thinking it should be possible to access Hypre AMS from deal.II via PETSc. I would like to play around with this idea. However, I have no experience in writing PETSc wrappers. Do you have a web page for a newbie PETSc wrapper or a link to a PR in which PETSc wrappers are discussed?

bangerth commented:
> Thanks! After saying my prayers I have copied the git magic words from the logbook into LCI. It looks like the squash-rebase was successful this time.

:-) Let's merge this as-is then. Everything else we can do in follow-up PRs if we want.

 
> I have another question. The PETSc has access to Hypre AMS. The deal.II has PETSc wrappers. I was thinking it should be possible to access Hypre AMS from deal.II via PETSc. I would like to play around with this idea. However, I have no experience in writing PETSc wrappers. Do you have a web page for a newbie PETSc wrapper or a link to a PR in which PETSc wrappers are discussed?

Yes, I did think about this as well when I saw the large number of iterations you need to allow in your linear solver. In principle, all you need to do is replace `SparseMatrix` by `PETScWrappers::MPI::SparseMatrix` and `Vector` by `PETScWrappers::MPI::Vector`. If you run the code on a single process, that's more or less everything you need to do, but because the constructors of the matrix and vector classes take different arguments, you also need ~20 lines of setup code to create some index sets. It's not a large amount of work, though. This then gets you to a state where you can run the code again with more or less the same functionality.

The next step would be to replace the preconditioner by `PETScWrappers::PreconditionBoomerAMG`, which is one of the hypre algorithms. We don't expose AMS at the moment, but I imagine adding the necessary functionality to `PreconditionBoomerAMG` (or creating a parallel class) is not going to be a lot of work -- PETSc makes that pretty easy, likely less than a day of work.

bangerth commented:
/rebuild

cembooks commented:
> Yes, I did think about this as well when I saw the large number of iterations you need to allow in your linear solver. In principle, all you need to do is replace `SparseMatrix` by `PETScWrappers::MPI::SparseMatrix` and `Vector` by `PETScWrappers::MPI::Vector`. If you run the code on a single process, that's more or less everything you need to do, but because the constructors of the matrix and vector classes take different arguments, you also need ~20 lines of setup code to create some index sets. It's not a large amount of work, though. This then gets you to a state where you can run the code again with more or less the same functionality.
> 
> The next step would be to replace the preconditioner by `PETScWrappers::PreconditionBoomerAMG`, which is one of the hypre algorithms. We don't expose AMS at the moment, but I imagine adding the necessary functionality to `PreconditionBoomerAMG` (or creating a parallel class) is not going to be a lot of work -- PETSc makes that pretty easy, likely less than a day of work.


Thanks! I will try to do that. I guess the best way to start is to write a PETSc-wrapper-based program with CG and then try to swap the CG for Hypre AMS.

bangerth commented:
> Thanks! I will try to do that. I guess the best way to start is to write a PETSc-wrapper-based program with CG and then try to swap the CG for Hypre AMS.

Yes, that sounds like the right approach.

bangerth commented:
Separately: Would you mind submitting a separate pull request in which you put an entry into `doc/news/changes/major` for your tutorial program?

cembooks commented:
> OK, it's merged! Welcome to the the deal.II authors!

Thanks!

cembooks commented:
> Separately: Would you mind submitting a separate pull request in which you put an entry into `doc/news/changes/major` for your tutorial program?


While doing this I have realized we have missed your name in the contributors list in intro.dox. Can we still correct this?

bangerth commented:
> While doing this I have realized we have missed your name in the contributors list in intro.dox. Can we still correct this?

No, that's alright. It's *your* program. I only reviewed it.
